# Importing libraries and datasets

In [1]:
import pandas as pd
import numpy as np
import os

In [6]:
path=r"C:\Users\Sam\Desktop\CF\Project 4 Instacart Basket\Instacart Basket Analysis\02 Data\Original Data"

In [12]:
df_prods=pd.read_csv(os.path.join(path,"products.csv"),index_col=False)

In [13]:
df_ords=pd.read_csv(os.path.join(path,"orders.csv"),index_col=False)

In [14]:
#create a test dataframe

In [16]:
df_test=pd.DataFrame()

In [17]:
#create a mix cloumn

In [18]:
df_test["mix"]=["a","b",1,True]

In [19]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [20]:
#function

In [23]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [24]:
df_test['mix'] = df_test['mix'].astype('str')

# Missing Values

In [26]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [32]:
df_nan=df_prods[df_prods["product_name"].isnull()==True]

In [33]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [35]:
df_prods.shape

(49693, 5)

In [39]:
df_prods_clean=df_prods[df_prods["product_name"].isnull()==False]

In [40]:
df_prods_clean.shape

(49677, 5)

# Duplicates

In [41]:
df_dups=df_prods_clean[df_prods_clean.duplicated()]

In [42]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [43]:
df_prods_clean.shape

(49677, 5)

In [44]:
df_prods_clean_nodups=df_prods_clean.drop_duplicates()

In [45]:
df_prods_clean_nodups.shape

(49672, 5)

In [46]:
path1=r"C:\Users\Sam\Desktop\CF\Project 4 Instacart Basket\Instacart Basket Analysis\02 Data\Prepared Data"

In [47]:
df_prods_clean_nodups.to_csv(os.path.join(path1,"products_checked.csv"))

# CLEANING DF_ORDS

## 2

In [48]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

There seems to be a lot of missing values in the days_since_prior_order varibale, since the count for the coulmn is significantly lower than the rest.

# 3

In [49]:
# checking for mixed datatypes in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)
else :
    print ("no mixed data type in the frame")

no mixed data type in the frame


# 5

In [50]:
# checking for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206209 missing values for the days_since_prior_order variable, this might be due to that many customers who only placed a single order, which will automatically put no value for them in the days_since_prior_order coulmn.

# 6 addresing the missing values

In [51]:
# creating a subset with only missing values
df_ords_nan=df_ords[df_ords["days_since_prior_order"].isnull()==True]


In [52]:
df_ords_nan

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [54]:
# checking for order number is the set and subset
df_ords["order_number"].value_counts()

order_number
1      206209
2      206209
3      206209
4      206209
5      182223
        ...  
96       1592
97       1525
98       1471
99       1421
100      1374
Name: count, Length: 100, dtype: int64

In [55]:
df_ords_nan["order_number"].value_counts()

order_number
1    206209
Name: count, dtype: int64

There are 206209 observations as the first order in the subset and in the set, which proves the point that there are 206209 missing values in the days_since_prior_order column for exactly that reason. Based on the this information, I think it would be the best to leave these values untouched.

# 7 duplicates

In [57]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [58]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

In [59]:
df_ords_dups.shape

(0, 7)

No duplicate values found

# 8 exporting cleaned data

In [60]:
# df_prods_clean_nodups exported in code line 47

In [62]:
# exporting df_ords
df_ords.to_csv(os.path.join(path1,"orders_checked.csv"))